### DISCUSSION TITLE
v4.0.33 pre-release

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#60] [General]
I've released v4.0.33 as a pre-release.

Many thanks to @isorin for his contributions.

To give this pre-release a go, please make sure that 'Show beta versions' is on when re-downloading from HACS.

As always, both positive and negative feedback is most welcome, and this will contribute to getting the release published as latest.

Note that until @BJReplay gets this Solcast integration published in HACS we're going to keep any releases as pre-release for now. That request progresses.

## What's Changed

* Performance improvements for sensor updates by @isorin, including:
  * Reduced the update interval of sensors to 5 minutes
  * Split the sensors into two groups: sensors that need to be updated every 5 minutes and sensors that need to be updated only when the data is refreshed or the date changes (daily values)
  * Fixed issues with removing the past forecasts (older than 2 years), broken code
  * Improve the functionality of the forecasts, for exmaple "forecast_remaining_today" is updated every 5 minutes by calculating the remaining energy from the current 30 minute interval. Same for "now/next hour" sensors.
* Redaction of Solcast API key in logs by @isorin
* Revert Oziee '4.0.23' async_update_options #54 by @autoSteve, which was causing dampening update issues

A comment from @isorin: "_I use the forecast_remaining_today to determine the time of the day when to start charging the batteries so that they will reach a predetermined charge in the evening. With my changes, this is possible._"

To that, I say nicely done.

New Contributors
* @isorin made their first contribution in https://github.com/BJReplay/ha-solcast-solar/pull/45

Full Changelog: https://github.com/BJReplay/ha-solcast-solar/compare/v4.0.32...v4.0.33

## Known issues

- The variable 'tally' should never be unavailable during a forecast fetch retry sequence, but it can be for some reason. This causes site 'forecast today' sensor to show as 'Unknown' until the retries are exhausted, or a successful fetch occurs.

- The call for get current API usage on startup is not correctly retried. This will be fixed in a future release.

- Startup times can be delayed when the Solcast API is returning 429/Too busy status, which is caused by retries. This will be fixed in a future release by getting the data immediately from cache if it exists.

**Comment by gcoan** [comment]
I've just installed v4.0.33 beta.

Checking the log afterwards there were some warnings about truncated future forecasts.  Don't know is this is significant ?

```

2024-06-26 00:17:05.488 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - API counter for ******Db75Av is 10/10
2024-06-26 00:17:05.524 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - load_saved_data file exists.. file type is <class 'dict'>
2024-06-26 00:17:05.551 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - removePastForecastData forecasts len in 1235
2024-06-26 00:17:05.552 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - removePastForecastData midnight_utc 2024-06-25 23:00:00+00:00, forecasts len out 324
2024-06-26 00:17:05.553 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-26 contains all 48 records
2024-06-26 00:17:05.553 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-27 contains all 48 records
2024-06-26 00:17:05.553 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-28 contains all 48 records
2024-06-26 00:17:05.553 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-29 contains all 48 records
2024-06-26 00:17:05.554 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-30 contains all 48 records
2024-06-26 00:17:05.554 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-07-01 contains all 48 records
2024-06-26 00:17:05.554 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-07-02 contains only 36 of 48 records and may produce inaccurate forecast data
2024-06-26 00:17:05.554 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-07-03 contains only 0 of 48 records and may produce inaccurate forecast data
2024-06-26 00:17:05.554 INFO (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - buildforecastdata processing took 0.0298s
2024-06-26 00:17:05.589 INFO (MainThread) [custom_components.solcast_solar] Solcast Integration version number: v4.0.33
2024-06-26 00:17:05.589 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Finished fetching solcast_solar data in 0.000 seconds (success: True)
```

The days it is warning about are day 7 and 8 from today, but there isn't a day 8 solcast forecast entity so 🤷‍♂️

![image](https://github.com/BJReplay/ha-solcast-solar/assets/142018870/8119099b-76da-4746-a260-a19c9788a504)
Day 7 shows a repeated drop to zero each night about midnight but I'm guessing this is because the daily sensors all shuffle one place forward and of course there is no day 8 so the forecast drops to zero before the next scheduled poll at 1am.
Whether this is right and it might be better preserving the prior forecast until day 7 is properly downloaded next time, might be worthwhile to prevent the zero's,  but I never look that far ahead at the forecast anyway

All the solcast sensors look OK to me, no unexpected values

**Reply by autoSteve** [reply]
AFAIK, @gcoan, the "truncations" are because the data is not yet in the response from Solcast. Perfectly normal.

Mine shows day eight as 0/48. Day seven 48/48. As one would expect, forecasting out that far is going to almost certainly be wrong anyway as the likelihood of hitting the average is low.

I'll defer to @isorin for comment regarding the drops to zero at midnight.

**Reply by autoSteve** [reply]
Just now on restart. Eight days fully returned all records, including today. All scheduled forecast gets are done for me for the day. It'd probably depend on time of day how much of day +7 can be had. It does interest me that six days out, your 2 July was a partial. It doesn't worry me, just interests.

Us Aussies are from the future, so maybe we can see further than you... 😂

```
SOLCAST - Data for 2024-06-26 contains all 48 records
SOLCAST - Data for 2024-06-27 contains all 48 records
SOLCAST - Data for 2024-06-28 contains all 48 records
SOLCAST - Data for 2024-06-29 contains all 48 records
SOLCAST - Data for 2024-06-30 contains all 48 records
SOLCAST - Data for 2024-07-01 contains all 48 records
SOLCAST - Data for 2024-07-02 contains all 48 records
SOLCAST - Data for 2024-07-03 contains all 48 records
```

**Reply by Noodleyman** [reply]
requesting the Aussies share tomorrows lottery numbers <3 🤣

**Reply by isorin** [reply]
@gcoan @autoSteve 
v4.0.33 should fix the issue with day 7 being truncated at midnight.
Before v4.0.33, we requested forecasts for exactly 168h (7 days) from the moment of update so we would have 7 full days plus a number of hours for the 8th day. For instance, if the update would run at 2pm then we would have forecasts for up to 2pm for the 8th day.
I've updated the logic to calculate the number of remaining hours for the 8th day so it would be complete at the time of update.
The reason for this is to avoid the truncation of the 7th day at midnight (when the 8th day becomes the 7th day) - and until the next scheduled update of data.
I also updated the method that checks the consistency of data to include all 8 days as, it only verified 6th days before.

@gcoan 
It is expected not to have the data in place for the 8th day until you do a scheduled update with v4.0.33 but I expected the 7th day to be complete. When it was your last update and can you also check to see if it reports all 48 records for 8 days after your 1st update?
Thanks.

**Reply by gcoan** [reply]
Solcast v4.0.33 been running all night.  There is a huge amount more in the log file (debug is turned on) from the 'per 5 minute' recalculation of solar remaining today, etc.

Scheduled rooftop polls worked OK

Interesting, the messages about missing records only appeared on Solcast startup, they don't appear when the update API is called - inconsistent?   I agree, I'm not worried about them, they are so far in the future, I just wonder whether they are needed, especially as day 8 doesn't exist in HA.

Only other warning message I see is the repeat of:

2024-06-26 06:15:12.392 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Solcast returned 373 records (should be 168)

Which has been happening ever since I turned debug on, but I did notice earlier in the log is the message:

2024-06-26 06:15:10.303 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Polling API for rooftop_id bc31-aaaa-bbbb-6a4e lastday 2024-07-03 23:00:00+00:00 numhours 186

Which probably explains the error, the API expects 186 hours of forecast but receives 373?  Solcast all appears to work fine, just these warning messages

**Reply by autoSteve** [reply]
> the messages about missing records only appeared on Solcast startup

The first start of v4.0.33 would have used cached data from 4.0.32...

**Reply by BJReplay** [reply]
> requesting the Aussies share tomorrows lottery numbers <3 🤣

My better half one first division more than quarter of a century ago. $10,000,000 prize.

Well, her and 100 colleagues shared a 1/10th share of it.

My prediction: This is the one and only time you'll win.  Quit while you're ahead.  Never buy another ticket!

**Reply by isorin** [reply]
@gcoan
The message "Solcast returned 373 records (should be 168)" it's fine, I think it was there before as well.
Even 168 was wrong before as we expected 168x2 (30 minutes intervals).
It's just a check and a message, nothing more but I will update it to reflect the reality. Thanks for noticing it!

The "debug" logging, that's true, more than it was before as it shows the calculations for the sensors that update every 5 minutes. I used the logging to verify my changes. Do we need to comment them out? I usually keep the logging level on "info" unless I need to investigate an issue,

Also, see my previous message about the 8th day.

**Reply by autoSteve** [reply]
> Quit while you're ahead

Pfft. You've got exactly the same chance winning tomorrow after a win today. Stuff all.

@Noodleyman, I'll PM you the winning numbers for your TZ. 😂

**Reply by gcoan** [reply]
Thanks for all your comments and to @isorin for the code improvements.

Checking the log in more detail, the 6:15 scheduled poll did indeed retrieve all records correctly, so the warning about missing records on restart must have been due to using the cached data from v4.0.32 as @autoSteve suggests:

```

2024-06-26 06:15:12.392 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Solcast returned 373 records (should be 168)
2024-06-26 06:15:12.411 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST http_data_call _fcasts_dict len 1295
2024-06-26 06:15:12.429 INFO (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - http_data_call processing took 0.037s
2024-06-26 06:15:12.454 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - removePastForecastData forecasts len in 1295
2024-06-26 06:15:12.455 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - removePastForecastData midnight_utc 2024-06-25 23:00:00+00:00, forecasts len out 384
2024-06-26 06:15:12.455 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-26 contains all 48 records
2024-06-26 06:15:12.456 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-27 contains all 48 records
2024-06-26 06:15:12.456 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-28 contains all 48 records
2024-06-26 06:15:12.456 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-29 contains all 48 records
2024-06-26 06:15:12.457 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-06-30 contains all 48 records
2024-06-26 06:15:12.457 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-07-01 contains all 48 records
2024-06-26 06:15:12.457 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-07-02 contains all 48 records
2024-06-26 06:15:12.457 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - Data for 2024-07-03 contains all 48 records
2024-06-26 06:15:12.458 INFO (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - buildforecastdata processing took 0.0273s
2024-06-26 06:15:12.502 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST get_total_kwh_forecast_day st 2024-06-25 23:00:00 end 2024-06-26 23:00:00 st_i 0 end_i 48 res 166.573
```

@isorin I'm not particularly anti the extra debug logging, after all it is what it is, debugging into, and normally wouldn't have the logging set that high, I was just commenting on it being present.

I'll check the reset to zero of d7 forecast at midnight tonight.

As for lottery numbers, if you promise not to tell anyone else, I'll share my foolproof lottery win strategy.  I've been doing it for years, by not buying a ticket each week I'm guaranteed to win £1 every time.  In fact since the price of the lottery ticket went up to £2 my weekly winnings have doubled.

**Comment by Noodleyman** [comment]
Thank you for the update, really appreciated. I'll be giving that a go now to check if it solves the problem I described today via https://github.com/BJReplay/ha-solcast-solar/discussions/28#discussioncomment-9878378

If I run into any issues I'll report back :)

**Reply by autoSteve** [reply]
If you don't run into issues please report back, @Noodleyman! This is a pre-release, after all. 😉

**Reply by Noodleyman** [reply]
Noted :), the issue was random in that I couldn't replicate it on demand, it would seem to randomly b0rk when updating the Dampening curves either via the UI or via the sevice call as it did last night. I think it's done it twice in a week using the service call to update, and I managed to trigger it once via the UI updating method. I've enabled my automation to update the values again which runs once per day at 00:03 to set the next days values to adjust for my shading in the mornings. 

Thank you for your efforts, it is very appreciated :)

**Comment by isorin** [comment]
@autoSteve 
Steve, thank you for v4.0.33 and for all the work you are doing around API calls, retries, etc.
It's good to see we keep this repo alive and glad to see my changes are getting some exposure as well.

**Reply by autoSteve** [reply]
A pleasure, @isorin.

But wait, there's more...

Another PR coming giving an even better retry experience, and redaction of personal information in debug logs like latitude and longitude, which essentially give out your home address.

This integration is already brilliant, and folks like you make it even more brilliant...er, so a heartfelt thanks to you.

It's my aim to make this integration the most stable and resilient that it can be, having witnessed the carnage of Oziee's reported issues list, a stack of which related to a lack of resiliancy.

**Comment by ProphetOfDoom** [comment]
Just updated to V4.033. As it's past midnight I checked and saw that my API count was set/reset to 0/10, as it should.

SW and NE roof tallys set ok and all other sensors populated.
No errors in log
Made a manual service call to solcast_solar.update_forecasts
API count now set to 2/10
SW and NE roof tallys set ok and all other sensors populated
No errors in log

As mentioned previously, my sole daily forecast update runs at sunrise - 7 mins (about 04:45), so I'll let you all know tomorrow how things go, but I've already bought some biscuits to celebrate!

As I'm not using the solcast forecast to actually do anything yet, I'll also try playing with the damping factors in the days to come.

**Comment by Noodleyman** [comment]
I'm not sure if this is a HA issue or a Solcast Solar issue, but sharing for visibility. 

Via HA -> Settings, there is a warning at the top of the page which reads as follows. the service is valid, but HA thinks it is not and flags it as a problem. (the Automation works fine).

![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/37700371-0c46-4e2a-9dcf-1af1a81c1ab5)

No issues with dampening data today, but my particular issue seemed random so needs time to review.

**Reply by autoSteve** [reply]
I take it you checked Developer Tools | Services to confirm that ```solcast_solar.set_dampening``` is available. And running that automation does set values. How odd.

No code has been changed in the BJReplay fork that relates to the dampening service that I'm aware of.

**Reply by Noodleyman** [reply]
Correct, working just fine.  I'm not overly worried about it as it is working correctly, just an odd error.
![image](https://github.com/BJReplay/ha-solcast-solar/assets/10614018/facb5e31-3bed-4f1e-b91d-dfa28b1f624d)

**Reply by autoSteve** [reply]
The below quotes are the closest that I can find doing a quick Google for similar. Sounds possibly transient.

Looking at the code, calling this service will restart the automation, and when that happens values briefly go unavailable. I have not confirmed this, but the service probably goes AWOL too until startup is complete.

Is it possible to dismiss the warning and see if it returns?

(I'm thinking about why an integration restart is required, and the answer may lie in dampening values being one of the 'options' for the integration. Changing other options, like forecast estimate also causes a restart. I have code for 4.0.next that improves restart time when the Solcast API is telling us to cool our jets with a 429/Too busy, because in that circumstance retries can slow down the restart. This delay, done at precisely the "right" time, could possibly be long enough for the repair center to flag it. And I don't think it gets automatically un-flagged should the service return.)

_"Hi everyone, I wanted to provide an update on the ‘unknown service’ warning I was experiencing. After some investigation, I discovered that this message was actually coming from the ‘repair center’ and not from the logs as I initially thought. Interestingly, when I clicked ‘dismiss’ on this error, it disappeared and hasn’t returned since. It’s possible that this warning was related to an outdated issue that has since been resolved. Thanks to all for your inputs and assistance!"_

and

_"It’s possible that the service is temporarily not available if the connection to Amazon fails intermittently. I don’t know if that’s a fact but it might be an explanation._

_"I use the Notifications for Android TV integration and if the app on the TV glitches out then HA says the service is missing. I just had that happen a few days ago."_

**Comment by Nilogax** [comment]
I have good news and bad news....

The good news is that I also have automations that use the solcast dampening service and have no warnings about the services being unknown.

The bad news is that despite working OK yesterday after I updated to 0.33, all my solcast data was zeroed this morning.  Log extract is at the bottom but to walk through the sequence of events the first step is an automation to recalculate dampening - at 01.06 I reset all dampening factors to 1 to give a raw forecast then at 01.14 I use the query_forecast_data service to capture historic forecasts so I can compare them against actual generation and calculate dampening.  The query_forecast_data service returned zeroes for historic forecasts, so my automation just reset the previous dampening factors.

Setting the dampening then resets the solcast integration and the API call to get the usage limit timed out so it tried to used cached data - I didn't have debug logging on (schoolboy error!) but as the generation history has gone I'm guessing that the cached data may also have been missing?

There are then a couple of scheduled calls to update the solcast forecast but they didn't return data, and the API usage count stayed at zero though the API last polled time was being updated.  

Once I noticed the issue this morning I restarted HA, and solcast then logged a bunch of errors (tally/rooftop ID, see log) but the zero forecast remained so I then did a manual call to update forecast and that seems to have resolved things. 

I'm guessing forecast history is lost as my last backup will be missing yesterday's figures.  If I can't restore the forecast history I'll need to disable the dampening update automations for a few days while it rebuilds enough historic forecast data to be useful for my calculations, so won't be able to re-test for a while.

Annotated log:
 
[solcast.log](https://github.com/user-attachments/files/16011342/solcast.log)

**Reply by autoSteve** [reply]
Can't bake a tasty omelette eithout breaking some eggs.

**First start:**

```SOLCAST - sites_data TimeoutError Error - Timed out connection to solcast server```, which is the first 500 I've seen, given usually 429. The sites cache will not have loaded, so things would turn to custard. This really wants fixing.

**Second start:**

What the actual (expletive) 🤯 ? On no line in this version is the rooftop site ID set to 'REDACTED' anywhere. The API key is redacted in logs, but not in stored variables. So how could this possibly be happening?

```
  File "/config/custom_components/solcast_solar/solcastapi.py", line 424, in get_rooftop_site_total_today
    if self._data["siteinfo"][rooftopid].get("tally") == None: _LOGGER.warning(f"SOLCAST - 'Tally' is currently unavailable for rooftop {rooftopid}")
       ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^
KeyError: 'REDACTED'
```

The only use of the word redacted has been this, in a log statement: ```_LOGGER.debug(f"SOLCAST - trying to connect to - {self.options.host}/rooftop_sites?format=json&api_key=REDACTED")```

**Reply by Nilogax** [reply]
That was me hiding the rooftop ID, not HA - is it useful?

**Reply by autoSteve** [reply]
Prolly no need to hide rooftop IDs. The only way you can get to that ID is with your API key (which should be redacted). I would be blown away if someone else's API key could get your rooftops, but what if they could? What evs. Knock yourself out with my forecast data.

What the log now says to me is that the intial integration load didn't contain any data for your rooftops. The file was maybe empty? Not sure yet. Was this because of the previous 500 error? Possibly, but I'll need to look at the sequence in code. (... Enters the Matrix...)

**Reply by autoSteve** [reply]
The timeout fascinates. This is not an HTTP error for sites data, rather the async function having a sixty second timeout. That could indicate a Solcast API issue that has returned 500/Internal server error, or a temporary Internet issue. If it were 500s then the retry would have worked and not hit that timeout. I'm leaning towards Internet.

When the log then moves on to the API usage get the last status is shown as 500/Internal server error, so now I'm tipping the former is the case, which is confusing. (The API usage code currently has an issue where it doesn't actually retry where it should, so the first call is getting 500. That retry is fixed in 4.0.next, BTW.)

In your log you then indicate 'FAILED FORECAST UPDATES'. There are no failures seen. This is happening in less than a 100/th the time it takes to blink an eye, so I'm tipping it's doing nothing. Probably because no rooftop sites from earlier.

Then restart.

Whoa.

```
2024-06-27 08:07:27.297 ERROR (MainThread) [custom_components.solcast_solar.solcastapi] SOLCAST - load_saved_data error: Traceback (most recent call last):
  File "/config/custom_components/solcast_solar/solcastapi.py", line 344, in load_saved_data
    await self.http_data_call(r_id=a, api=ks[a], dopast=True)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: SolcastApi.http_data_call() missing 1 required positional argument: 'usageCacheFileName'
```

**This is a bona fide bug**. This is trying to get rooftop history and the call is just wrong. But there is a comment in code at this point...

```some api keys rooftop data does not exist yet so go and get it```

What has just happened is that the rooftop sites cache has been loaded, and then is being inspected for whether each site has cached data. If debug logging were turned on it would report a sickening ```_LOGGER.debug("SOLCAST - Must be new API jey [sic] added so go and get the data for it")```. This is not sickening because of the cringe-worthy jey typo, but rather that it does not have the data. Has your ```/homeassistant/solcast.json``` sites cache got stuff in it?

The rest is kind of expected after this cluster. But actually better by accident in current 4.0.next.

**Reply by Nilogax** [reply]
Yes solcast.json is populated now, as are solcast-sites and -usage.   The forecast history was already gone before the 500 timeout, so is that potentially a separate issue with 0.33?

**Reply by autoSteve** [reply]
Possibly. Difficult to say without the prior log, which is probably gone now because restarts.

You have uncovered lots in tracing this.

**Reply by autoSteve** [reply]
4.0.next should cover:

* Sites get timeout will fall back to cache if it exists
* Oziee typos, and MANY more logging improvements
* Sites cache being sometimes incorrectly created with the API key appended as solcast-sites-[apikeyhere].json, despite only having one API key
* Redaction of lat/longitude in debug logs
* Fix dodgy get intial data in 4.0.33 should solcast.json not exist
* Get API usage on call of remove solcast.json cache / essentially a reset
* Elimination of 'tally' warnings most likely
* Fix API usage retry mechanism
* Instantly fall-back to cache on reload if rooftop/usage API calls fail

**Reply by gcoan** [reply]
> Sites cache being sometimes incorrectly created with the API key appended as solcast-sites-[apikeyhere].json, despite only having one API key

I've seen this, I have a solcast-sites-key.json and a solcast-sites.json.  I wondered about it, it's a bug then.  Can I delete the extra file?

**Reply by autoSteve** [reply]
Yep, @gcoan. PR https://github.com/BJReplay/ha-solcast-solar/pull/61 for the 'alpha' release brave... 😉 Just solcastapi.py. It's running on my 'prod' 🤞, but has not even a day under its belt yet...

**Reply by gcoan** [reply]
> Yep, @gcoan. PR https://github.com/BJReplay/ha-solcast-solar/pull/61 for the 'alpha' release brave... 😉 Just solcastapi.py. It's running on my 'prod' 🤞, but has not even a day under its belt yet...

Not brave.  I'll wait a bit thanks @autoSteve

**Reply by autoSteve** [reply]
I love being a crash-test dummy for you lot...

**Comment by gcoan** [comment]
you break it, you fix it….On 27 Jun 2024, at 14:50, Steve Saunders ***@***.***> wrote:﻿
I love being a crash-test dummy for you lot...

**Reply by autoSteve** [reply]
In the words of Uncle Roger (see Youtube), "So weak, so weak, you can not be along for the ride?" 🤣

**Comment by isorin** [comment]
@Nilogax 
I think this service call: "query_forecast_data" is affected by the performance improvements in v4.0.33 - it won't return any forecasts from the past.
The original data is there, in another structure, I'll try to come up with a fix for it.

This shouldn't affect history (persisted or otherwise).

**Reply by Nilogax** [reply]
Ah OK. Let me know if you want me to play crash test dummy on any changes. Can't let @autoSteve have all the fun...

**Reply by isorin** [reply]
I've never played with dampening, how do you use it?
It looks (from reading your comments) that you have some advanced logic built around it.
How far do you go back when recalculating the factors?

**Reply by Nilogax** [reply]
I use dampening to reflect the shading on my array so that predbat has more representative data to work from. As the timing and severity of the shading varies over time I decided to automate the dampening settings, and with minimal knowledge of HA or Python and a bit of Googling I cobbled something together. I wouldn't go as far as calling it advanced logic though!

I settled on comparing the last 5 days of forecast vs actual generation as that seems to be enough to smooth the effect of good/bad weather while still being fairly responsive to the changes in shading.

**Reply by isorin** [reply]
That's interesting, I might look into it as I have 4 panels on south-west that are affected by shading (for me it depends on the elevation of the sun)

Anyway, PR is up with fix for query_forecast_data:
https://github.com/BJReplay/ha-solcast-solar/pull/62

Once @autoSteve and/or @BJReplay merge the changes into a new pre-release we can test them.

@Nilogax Actually, you can grab this file (as all changes are in one file)
https://github.com/BJReplay/ha-solcast-solar/blob/8a5cc8a90a11794941c336fa2e1560942e84e60d/custom_components/solcast_solar/solcastapi.py
override the same file in HA (config/custom_components/solcast_solar/solcastapi.py) and try it (if you want/can - if not we can wait for a pre-release).

I tried it on mine:
```
service: solcast_solar.query_forecast_data
data:
  start_date_time: "2024-06-25 13:00:00"
  end_date_time: "2024-06-25 16:00:00"

Response:
data:
  - period_start: "2024-06-25T13:00:00+01:00"
    pv_estimate: 2.0619
    pv_estimate10: 1.1329
    pv_estimate90: 3.423
  - period_start: "2024-06-25T13:30:00+01:00"
    pv_estimate: 4.8344
    pv_estimate10: 4.0485
    pv_estimate90: 4.9306
  - period_start: "2024-06-25T14:00:00+01:00"
    pv_estimate: 4.893
    pv_estimate10: 4.0773
    pv_estimate90: 4.893
  - period_start: "2024-06-25T14:30:00+01:00"
    pv_estimate: 4.3993
    pv_estimate10: 2.6204
    pv_estimate90: 4.7671
  - period_start: "2024-06-25T15:00:00+01:00"
    pv_estimate: 3.5179
    pv_estimate10: 2.0306
    pv_estimate90: 4.6016
  - period_start: "2024-06-25T15:30:00+01:00"
    pv_estimate: 2.5289
    pv_estimate10: 1.7617
    pv_estimate90: 3.376
```

**Reply by Nilogax** [reply]
Thanks for the quick work @isorin.  I've replaced solcastapi.py and can now access yesterday's forecast, but any data prior to my installation of 0.33 is gone (LTS had already gone but I still had the recent forecast history).  I'll pause my dampening stuff while the history rebuilds but will do some manual checks over the coming days too, otherwise no errors or warnings in the log from 0.33.

See https://github.com/BJReplay/ha-solcast-solar/discussions/28 if you want to discuss dampening further

**Reply by autoSteve** [reply]
I'm running @isorin's proposed changes, in conjunction with mine for 4.0.next and I still see the recent forecast history, @Nilogax. It's early for me because I only recently locally incorporated all proposals.

The loss of LTS because of a prior change in state_class for that sensor is unfortunate, but that history data is of almost zero value. You should still have the recent past, though.

@isorin, I take it you still have recent history for the month?

**Reply by Nilogax** [reply]
@autoSteve Yeah I'm not fussed about the LTS loss. I'm thinking the 500 error is possibly what nuked my history, as the earlier zero response to query forecast data would have been expected from @isorin's changes.

**Reply by isorin** [reply]
@autoSteve
I checked and my history goes as far back as "2024-06-09T23:30:00+00:00"
